### PR TITLE
Add a blog post on the design of versioned-hdf5

### PR DESCRIPTION
**Pull request by asmeurer**
Some of this was copied or modified from the versioned-hdf5 design
documentation at https://deshaw.github.io/versioned-hdf5/design.html. Unlike
that document, I have omitted some of the more technical implementation
details, and focused more on details here on how copy-on-write works.

### CODE REVIEW

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+abstraction for HDF5 groups and datasets. The basic idea behind the design is
+that versioned-hdf5 is a
+[copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write) system. The basic
+idea of copy-on-write is that all data is effectively immutable in the
+backend. Whenever a high-level representation of data is modified, it is
+copied to a new location in the backend, leaving the original version intact.
+Any references to the original will continue to point to it.
+
+The basic high-level idea of a versioned system built on top of copy-on-write
+is that is stored in the system is stored in immutable blobs in the backend.
```
Comment (melissawm) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:29]: I think this sentence has some extra words

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+of their contents. Cryptographic hashes form a mapping from any blob of data
+to a fixed set of bytes, which is effectively one-to-one, meaning if two blobs
+have the same hash, they must be exactly the same data. This means that it is
+impossible to mutate a blob of data in-place. Doing so would change its hash,
+which would make it a different blob, since blobs are referenced only by their
+hash.
+
+Whenever data for a version is committed, its data is stored as blobs in the
+backend. It may be put into a single blob, or split into multiple blobs. If it
+is split, a way to reoncstruct the data from the blobs is stored. If a later
```
Comment (melissawm) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:42]: Typo in "reoncstruct" :)

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+the resulting versioned file unnecessarily large, as changing even a single
+element of a chunk requires rewriting the entire chunk. Versioned-hdf5 does
+not presently contain any logic for automatically picking a chunk size. The
+[pytables
+documentation](https://www.pytables.org/usersguide/optimization.html) has some
+tips on picking an optimal chunk size.
+
+Because chunks are the most basic HDF5 primitive, Versioned-HDF5 uses them as
+the underlying blobs for storage. This way operations on data can be as
+performant possible.
```
Comment (melissawm) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:76]: I think the last sentence should read "as performant **as** possible".

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+[Virtual datasets](http://docs.h5py.org/en/stable/vds.html) are a special kind
+of dataset that reference data from other datasets in a seamless way. The data
+from each part of a virtual dataset comes from another dataset. HDF5 does this
+seamlessly, so that a virtual dataset appears to be a normal dataset.
+
+The basic design of versioned-hdf5 is this: whenever a dataset is created for
+the first time (the first version containing the dataset), it is split into
+chunks. The data in each chunk is hashed and stored in a hash table. The
+unique chunks are then appended into to a `raw_data` dataset corresponding to
```
Comment (melissawm) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:87]: "into to a `raw_data`" -> "into a `raw_data`"

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+-->
+
+In a previous post, we introduced the Versioned HDF5 library and described
+some of its features. In this post, we'll go into detail on how the underlying
+design of the library.
+
+<!-- TEASER_END -->
+
+The Versioned-HDF5 library is a library that wraps h5py and offers a versioned
```
[the remark this answers is not in the document]
Comment (asmeurer) [reply posts/2020/09/design-of-the-versioned-hdf5-library.markdown:19]: Oh I think the dash should not be there if we are capitalizing it like this. Either `versioned_hdf5` for the Python library name, `versioned-hdf5` for the package name, or `Versioned HDF5` in plain text. I haven't been very consistent with this.

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+.. date: 2020-08-31 16:30:26 UTC-05:00
+.. tags:
+.. category:
+.. link:
+.. description:
+.. type: text
+
+-->
+
+In a previous post, we introduced the Versioned HDF5 library and described
```
Comment (scopatz) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:13]: Please add link to the previous post

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown (hunk 5 above)
[the remark this answers is not in the document]
Comment (scopatz) [reply posts/2020/09/design-of-the-versioned-hdf5-library.markdown:19]: Also, probably could be "Versioned HDF5 is a library that ..."

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+some of its features. In this post, we'll go into detail on how the underlying
+design of the library.
+
+<!-- TEASER_END -->
+
+The Versioned-HDF5 library is a library that wraps h5py and offers a versioned
+abstraction for HDF5 groups and datasets. The basic idea behind the design is
+that versioned-hdf5 is a
+[copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write) system. The basic
+idea of copy-on-write is that all data is effectively immutable in the
```
Comment (scopatz) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:23]: Probably should rephrase to not use `The basic idea` two sentences in a row for different concepts

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+backend. Whenever a high-level representation of data is modified, it is
+copied to a new location in the backend, leaving the original version intact.
+Any references to the original will continue to point to it.
+
+The basic high-level idea of a versioned system built on top of copy-on-write
```
Comment (scopatz) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:28]: `Basic idea` again

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+hash.
+
+Whenever data for a version is committed, its data is stored as blobs in the
+backend. It may be put into a single blob, or split into multiple blobs. If it
+is split, a way to reoncstruct the data from the blobs is stored. If a later
+version modifies that data, any blobs that are different are stored as new
+blobs. If the data is the same, the blobs will also be the same, and hence
+will not be written to twice, because they will already exist in the backend.
+
+At a high-level, this is how the git version control system works, for
```
Comment (scopatz) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:47]: link to git book?

File: posts/2020/09/design-of-the-versioned-hdf5-library.markdown [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+virtual datasets in the corresponding version groups in `versions/`. For
+example, the chunks for the data `data1` in versions `version1` and `version2`
+are stored in `_versioned_data/data1/raw_data`.
+
+## Versioned-HDF5 API
+
+The biggest challenge of this design is that the virtual datasets representing
+the data in each versioned data all point to the same blobs in the backend.
+However, in HDF5, if a virtual dataset is written to, it will write to the
+location it points to. This is at ends with the immutable copy-on-write
```
Comment (scopatz) [inline posts/2020/09/design-of-the-versioned-hdf5-library.markdown:166]: at odds?

**Comment by asmeurer** [comment]
I've addressed all the comments here.

**Comment by asmeurer** [comment]
My post links to [Melissa's](https://github.com/Quansight-Labs/quansight-labs-site/pull/118), so you may want to merge that one first. I don't know if we want to spread the posts out so they aren't both posted at the same time.

**Comment by rgommers** [comment]
> My post links to [Melissa's](https://github.com/Quansight-Labs/quansight-labs-site/pull/118), so you may want to merge that one first. I don't know if we want to spread the posts out so they aren't both posted at the same time.

Yes, that'd be better. 4-7 days apart I think.

**Review by melissawm — APPROVED** [review]

**Comment by rgommers** [comment]
Post 2 was published, so aiming for next Tue/Wed for this one?

**Comment by rgommers** [comment]
Nice post, thanks @asmeurer! In it goes.

**Pull request merged**